pkg/controller/pkg/revision/hook.go

```diff
@@ -0,0 +1,178 @@
+}
+
+// ProviderHook preforms operations for a provider package that requires a
+// controller before and after the revision establishes objects.
+type ProviderHook struct {
+	client    resource.ClientApplicator
+	namespace string
+}
+
+// NewProviderHook creates a new ProviderHook.
+func NewProviderHook(client resource.ClientApplicator, namespace string) *ProviderHook {
+	return &ProviderHook{
+		client:    client,
+		namespace: namespace,
+	}
+}
+
+// Pre cleans up a packaged controller and service account if the
+// revision is inactive.
+func (h *ProviderHook) Pre(ctx context.Context, pkg runtime.Object, pr v1alpha1.PackageRevision) error {
+	pkgProvider, ok := pkg.(*pkgmeta.Provider)
+	if !ok {
+		return errors.New(errNotProvider)
+	}
+	// Always set revision status fields.
+	pr.SetDependencies(convertDependencies(pkgProvider.Spec.DependsOn))
+	pr.SetCrossplaneVersion(reference.FromPtrValue(pkgProvider.Spec.Crossplane))
+
+	// Do not clean up SA and controller if revision is not inactive.
+	if pr.GetDesiredState() != v1alpha1.PackageRevisionInactive {
+		return nil
+	}
+	s, d := buildProviderDeployment(pkgProvider, pr, h.namespace)
+	if err := h.client.Delete(ctx, d); resource.IgnoreNotFound(err) != nil {
+		return errors.Wrap(err, errDeleteProviderDeployment)
+	}
+	if err := h.client.Delete(ctx, s); resource.IgnoreNotFound(err) != nil {
+		return errors.Wrap(err, errDeleteProviderSA)
+	}
+	return nil
+}
+
+// Post creates a packaged provider controller and service account if the
+// revision is active.
+func (h *ProviderHook) Post(ctx context.Context, pkg runtime.Object, pr v1alpha1.PackageRevision) error {
+	pkgProvider, ok := pkg.(*pkgmeta.Provider)
+	if !ok {
+		return errors.New("not a provider package")
+	}
+	if pr.GetDesiredState() != v1alpha1.PackageRevisionActive {
+		return nil
+	}
+	s, d := buildProviderDeployment(pkgProvider, pr, h.namespace)
+	if err := h.client.Apply(ctx, s); err != nil {
+		return errors.Wrap(err, errApplyProviderSA)
+	}
+	if err := h.client.Apply(ctx, d); err != nil {
+		return errors.Wrap(err, errApplyProviderDeployment)
+	}
+	pr.SetControllerReference(runtimev1alpha1.Reference{Name: d.GetName()})
+	return nil
+}
+
+// ConfigurationHook preforms operations for a configuration package before and
+// after the revision establishes objects.
+type ConfigurationHook struct{}
+
+// NewConfigurationHook creates a new ConfigurationHook.
+func NewConfigurationHook() *ConfigurationHook {
+	return &ConfigurationHook{}
+}
+
+// Pre sets status fields based on the configuration package.
+func (h *ConfigurationHook) Pre(ctx context.Context, pkg runtime.Object, pr v1alpha1.PackageRevision) error {
+	pkgConfig, ok := pkg.(*pkgmeta.Configuration)
+	if !ok {
+		return errors.New(errNotConfiguration)
+	}
+	// Always set revision status fields.
+	pr.SetDependencies(convertDependencies(pkgConfig.Spec.DependsOn))
+	pr.SetCrossplaneVersion(reference.FromPtrValue(pkgConfig.Spec.Crossplane))
+	return nil
+}
```

Comment:
This now basically does a subset of what the `ProviderHook` does in its `Pre` step (minus the type assertion). I thought about making a common interface and passing a set of hooks to each controller (i.e. Configuration controller would just take a `StatusSetter` hook, while Provider controller could take a `StatusSetter` and a `ControllerBuilder`). However, given that things like `PermissionRequests` may be reintroduced in the future, and we may have more package types, just having separate, self-contained pre and post hooks for each package type seemed like the best option for now.